pkg/webhook/validator.go

```diff
@@ -0,0 +1,108 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package webhook
+
+import (
+	"context"
+	"net/http"
+
+	"k8s.io/api/admission/v1beta1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/runtime/inject"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+// Validator defines functions for validating an operation
+type Validator interface {
+	runtime.Object
+	ValidateCreate() error
+	ValidateUpdate(old runtime.Object) error
+}
+
+// newValidatingWebhookFor creates a new admission.Webhook for Validating the provided type.
+// Returns nil if o does not implement the Validator interface.
+func newValidatingWebhookFor(o runtime.Object) *admission.Webhook {
+	v, ok := o.(Validator)
+	if !ok {
+		// Type doesn't implement function for validating webhook
+		return nil
+	}
+	return &admission.Webhook{
+		Handler: &validatingHandler{validator: v},
+	}
+}
+
+type validatingHandler struct {
+	validator Validator
+	scheme    *runtime.Scheme
+	decoder   *admission.Decoder
+}
+
+var _ inject.Scheme = &validatingHandler{}
+
+// InjectDecoder injects the decoder into the FirstMateDeleteHandler
+func (h *validatingHandler) InjectScheme(s *runtime.Scheme) error {
+	h.scheme = s
+	var err error
+	h.decoder, err = admission.NewDecoder(h.scheme)
```

Comment:
decoders should automatically be injected with `InjectDecoder` by the `admission.Webhook` logic.